src/FuncSharp/Extensions/IEnumerableExtensions.tt

```diff
@@ -87,19 +87,55 @@ namespace FuncSharp
         {
             return source.SingleOption().OrElse(_ => source.FirstOption().Map<Exception>(e => new AggregateException(source)));
         }
+        <#  for (var i = 1; i <= MaxCubeArity(); i++) { #>
+
+        /// <summary>
+        /// Converts the source to a new <#= i #>-dimensional data cube and aggregates values.
+        /// </summary>
+        public static <#= DataCubeType(i, valueType: "TResult") #> ToDataCube<T, <#= Types(i, name: "P") #>, TValue, TResult>(
+            this IEnumerable<T> source,
+<#= Lines(i, x => Indent(12) + "Func<T, P" + x + "> p" + x, separator: ",") #>,
+            Func<T, TValue> value,
+            Func<TValue, TResult> initialization,
+            Func<TResult, TResult, TResult> aggregation)
```

Comment:
The aggregation method should be of type `Func<TResult, TValue, TResult>` to avoid possibly useless initialization which is visible here https://github.com/siroky/FuncSharp/pull/92/files#diff-c66b6a7351d5ead614565a2b81008b44155ec1dd083fa74cef8d3e7421311495R368 with the Enumerable.Repeat. The initialization should be used only first time for the first element. And all the others should be aggregated without the need of initialization.

Maybe this will require another variant of SetOrElseUpdate, but that could be helpful as well.

Reply:
@siroky there is still some initialization as current C# in funcsharp doesnt have append yet.
https://github.com/siroky/FuncSharp/pull/92/files#diff-c66b6a7351d5ead614565a2b81008b44155ec1dd083fa74cef8d3e7421311495R297